File: NinePSharp.Messages.FSharp/messages/Trenameat.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Trenameat =
    { Size: uint32
      Tag: uint16
      OldDirFid: uint32
      OldName: string
      NewDirFid: uint32
      NewName: string }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Trenameat
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Trenameat span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.OldDirFid); offset <- offset + 4
            writeString this.OldName span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.NewDirFid); offset <- offset + 4
            writeString this.NewName span &offset

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let oldDirFid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let oldName = readString data &offset
        let newDirFid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let newName = readString data &offset
        { Size = size; Tag = tag; OldDirFid = oldDirFid; OldName = oldName; NewDirFid = newDirFid; NewName = newName }

File: NinePSharp/messages/Trenameat.cs
namespace NinePSharp.Messages;

using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

public readonly struct Trenameat : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Trenameat;
    public ushort Tag { get; }
    public uint OldDirFid { get; }
    public string OldName { get; }
    public uint NewDirFid { get; }
    public string NewName { get; }

    public Trenameat(uint size, ushort tag, uint oldDirFid, string oldName, uint newDirFid, string newName)
    {
        Size = size;
        Tag = tag;
        OldDirFid = oldDirFid;
        OldName = oldName;
        NewDirFid = newDirFid;
        NewName = newName;
    }

    public Trenameat(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        int offset = NinePConstants.HeaderSize;
        OldDirFid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        OldName = data.ReadString(ref offset);
        NewDirFid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        NewName = data.ReadString(ref offset);
    }

    public void WriteTo(Span<byte> span)
    {
        span.WriteHeaders(Size, Tag, MessageTypes.Trenameat);
        int offset = NinePConstants.HeaderSize;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), OldDirFid); offset += 4;
        span.WriteString(OldName, ref offset);
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), NewDirFid); offset += 4;
        span.WriteString(NewName, ref offset);
    }
}
